### PR TITLE
Add new syntax for Java actions with Http.Context

### LINKED ISSUE #6168: Figure out how to remove Java thread local Context
Java `Http.Context.current()` uses a thread local. Play 2.6 should provide an easy way to migrate to something that does not use thread local state.

One solution might be to make Java method calls look like:

``` java
public Result foo(Context cxt, String param1, int param2) {
  // ...
}
```

If we did that, we'd need to decide if there should be a special syntax in the routes for adding the context.

Another idea is

``` java
public FunctionalResult foo(String param1, int param2) {
  return ctx -> {
    // ...
  }
}
```

That might be a starting point for the evolution toward a more functional API.

I see this as a more important goal than creating a completely new API for Java actions, which is something that could potentially wait for 3.0.

### PR DESCRIPTION
**Pull request by gmethvin**
Fixes #6168 

Introduces a new syntax for Java actions that does not require statics, for example:

``` java
public Function<Context, Result> index() {
  return ctx -> {
    ctx.response().setHeader("Foo", "bar");
    return ok("Hello world!");
  };
}
```

### CODE REVIEW

File: documentation/manual/working/javaGuide/main/http/code/javaguide/http/JavaActions.java [hunk 1]
```diff
@@ -69,6 +71,34 @@ public Result index() {
...
+                };
+            }
+            //#result-context
+        }, fakeRequest(), mat).status(), equalTo(200));
+    }
+
+    @Test
+    public void simpleResultWithContextAsync() {
+        assertThat(call(new MockJavaAction() {
+            //#simple-result-context-async
```
Comment (zhangshifeng) [inline documentation/manual/working/javaGuide/main/http/code/javaguide/http/JavaActions.java]: Label name don't match.

Comment (gmethvin) [inline documentation/manual/working/javaGuide/main/http/code/javaguide/http/JavaActions.java]: thanks

**Comment by gmethvin** [comment]
I've also refactored the Java `Action` interface, so now it's possible to wrap actions in other actions, like this:
```java
public AsyncHandler index() {
  return csrfAction.wrap(ctx -> {
    ctx.response().setHeader("Foo", "bar");
    return ok("Hello world!");
  });
}
```

It does create a small breaking change if you're defining an action: instead of defining `call(Context context)` you need to define `call(Context ctx, AsyncHandler delegate)`. The difference is essentially that `delegate` is passed in rather than being a field. For callers, the original `call(Context ctx)` delegates to this one so no changes are needed. This change allows us to wrap other actions without mutating the original action.

I'm not sure about the naming of `AsyncHandler` and `ActionHandler` though. Suggestions for better names welcome.

**Comment by richdougherty** [comment]
What do you think about making the API even more like the Scala one? i.e. Use functional style only, drop support for annotations, etc.

```java
import play.mvc.Action;
import play.mvc.ActionBuilder;
import play.mvc.Controller;

@Inject
public class HomeController(action: ActionBuilder) extends Controller {
  Action index() {
    return action.sync(ctx -> Results.ok("Hello"));
  }
}
```

**Comment by gmethvin** [comment]
@richdougherty That would be the second phase of this. I think all we need is to make a builder that returns some type of `EssentialAction`. The goal for this PR is to still allow using the old action style without having to use statics for `Http.Context`.

**Comment by richdougherty** [comment]
Right—I was thinking it might be easier to skip the first phase and just do the second phase! :)

Also, if we're introducing functional style for making Java actions, I think it might be better if we stick to function composition only for composing actions with lambdas and drop annotation support when composing lambda actions. Two reasons for that: 1. the code is probably easier for you to write, 2. I think it would make the transition to our new API easier for users to understand if we have an old method/annotation API for actions and a new functional API—if we also have a hybrid between the two then it could be a bit confusing.

**Comment by gmethvin** [comment]
@richdougherty I think there is some value in being able to mix the two, for people who want an explicit way to get access to the context but don't want to refactor all their code.

I also wonder if for the pure functional approach we should remove `Context` completely and move that functionality directly to `play.mvc.RequestHeader`.

**Comment by tajakobsen** [comment]
Have you considered creating interfaces that extends `Function<Context, Result>` and `Function<Context, CompletionStage<Result>>` to give more readable method declarations in Controllers?

```java
public interface RenameMe1 extends Function<Context, Result> {}
```
```java
public interface RenameMe2 extends Function<Context, CompletionStage<Result>> {}
```

**Comment by gmethvin** [comment]
@tajakobsen Actually that is what I had originally, but @richdougherty and I discussed and decided `Function` would be a better choice for the actual API. I don't see a reason to force people to use a custom type. It will work with any type that extends `Function<Context, CompletionStage<Result>>` though.

**Comment by mkurz** [comment]
I mentioned it in #6171 already, but maybe it fits here as well:
We should also pass the current `Http.Context` to the various `onError` methods in the `HttpErrorHandler`.
We already had cases where we used `Http.Context.current()` in our ErrorHandlers.

**Comment by gmethvin** [comment]
@mkurz This PR is still open since we're not actually sure if we want to go in this direction. We're thinking about a new API that doesn't use `Context` and just uses a new type of request, much more like the current Scala API. That way you don't need any of the Java "magic" enhancement and can unit test your controllers by just passing something to the method.

We already have the foundations for this in the Java `play.mvc.EssentialAction` and actually it's totally valid to have a Java controller that has methods returning `EssentialAction`. We just need to design a nice functional API that allows you to construct `EssentialActions` and access everything directly from the `Request` like you can in Scala. What do you think?

**Pull request closed without merging**